### PR TITLE
Transfer learning with VGG19 using torchvision

### PR DESCRIPTION
**Pull request by sitamgithub-MSIT**
This notebook applies transfer learning with the "ConvNet as fixed feature extractor" approach. vgg19 is the model used for the base-pretrained model.

## Additional Changes Made
- Added remarks to several different code sections.
- Rather than placing everything in one function, distinct helper functions were implemented for training.
- Improved interpretibility of functions by assigning corresponding document strings.
Additionally, data augmentation and processing were addressed.

cc: @merveenoyan, @MKhalusova

### CODE REVIEW

**Comment by sitamgithub-MSIT** [comment]
I faced a problem while using Plotly in which the image needed to be static to display inside the notebook. To resolve this, I used a picture of my graph and created a pull request for it on huggingface-cv-org. I attempted to use other contributor images just for testing, but they were not rendering properly inside the notebook. As a solution, I placed the picture inside my working unit under the notebook folder.

**Comment by sitamgithub-MSIT** [comment]
> I faced a problem while using Plotly in which the image needed to be static to display inside the notebook. To resolve this, I used a picture of my graph and created a pull request for it on huggingface-cv-org. I attempted to use other contributor images just for testing, but they were not rendering properly inside the notebook. As a solution, I placed the picture inside my working unit under the notebook folder.

Though in ReviewNB, the plotly graph is visible.

File: notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb [hunk 1]
```diff
@@ -0,0 +1 @@
...VF45b3orRzLAeTvFj5TzugaIG4Wn84dxDbtRUgynP4hsa9Bq7jtYyrWwuv1VTaT8G7E60M6ZwAF8DYQUKHwID4uAQkRIjJ2wCJ7BYBb/O6q3x5ZNKXLimTDikgHYvjpoOwSU01kQAQGS4kMfVsCrB2GoFMIZRGjJOx4FqIGYpLYO/A2sfWQn/45VN45KSsK8/5tjmWyx5t/0cS9au/CgPWID4NyfESEfiP4NoE10ZQE+G6CLiAAQGRAxIPCBzhaQXnV0h+wOASsHCgaGam0obOJXiOCMOAdb/GarXCznoHq6HHEKN47jkH8m2JDcLG41TN3+aD7QDv4LzX8OBe+N6CStja4QjE6p+VrPVlgWOFnJs6c/uYGaPzgG8InXNIvgV1jDYCfWwQB8aQJNlzNK2sgR/24ERoHdA0DdqmwWLGGLaAbtFivuiwWHaYLzrM5o10XQQoMXwAll2L5YE5ZvMWz/PA6v4I9BEOwMx7LGct9i0XWPcBQ4hY9S0a1+d1JaWo5skO/dBj3TP8OklIbp8wxBYtvGh9qEELyR3UokVHLRy3os0EQSJIEvi0Q9xm9H/IWO0krFYJO6sBvYKf1aqBxG0/Nzov8PPbv/3bo++/9mu/hksvvRR33XUXvv/7vx/Hjx/Hv//3/x7vfOc78Wf/7J8FALzjHe/AC17wAnzkIx/By1/+cvyP//E/8LnPfQ7/83/+Txw5cgTf9V3fhV/4hV/Am9/8Zvzjf/yP0XXnGbJhT6qBCOnkKQtDVkhy8S+waTnfXQlVtfYmQQQr9dWETScZXFXvtwWs1g5R9f+ZosbsrgmZfFnkPwC2hpCaK4kHHUDkwG2jDr7SBs479L3HEAIGGiSCSgAYDjEx+iFh1cdS8koOYhbtUAzifByGhNV6QL8OGIaY7cHHwKdUqF5UyulNAFR3xLhlaNfB3u00ehrv9X3z5B73mZy7lww4XvnLuRLCaYTfxHHWUm7yqFYELr44RKPHZj8iXXzzzmdVXuGrIpCTCjxiSsW6IwdxbK3LBDODQbYVTswbwrwsUBncqL9LnUU8RybMxS+CvvCqPUcFUTWzkHpXHaALp9S7Hh11g5TLszM9tE91oBbfKwi4soXLUemfrJ4y0cfa2DrODOGKwGBCey4rKiLoHOPAWehLecwV4ENZ6rCcYdZnNqds8lstNMm4t++jgQnsYuLq9+p7ve
```
Comment (merveenoyan) [inline notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: <p>I think it would be better if the label and the image were side by side, also would be nice to load the image from course-assets repository</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='38'/>

**Review by merveenoyan — COMMENTED** [review]
Overall content is very good, also liked that you have put explanations on top. I think what you can do to improve it even more is that you can also note down the motivation behind helper functions and maybe explain step by step what you can do and also carry some of the comments on the markdown above the cells.

File: notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb (hunk 1 above)
Comment (MKhalusova) [inline notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: <p>My personal preference is to have imports in cells where they are first used in educational notebooks. This way, someone not familiar with some of the libraries or acronyms (e.g. <code>plotly.graph_objects as go</code>)will not have to scroll up and down, and back up again to look up what it is that you're using. Your way is closer to what one may see in production, but it may be harder to follow for learners.  This is not wrong, of course, and, like I said, it's a personal preference, so I'll leave it up to you.   </p><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (MKhalusova) [inline notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: <p>Why not load the data from the Hugging Face Hub with the datasets library? </p><p> </p><p>Nit: <code>The data is available in 🤗 Face Datasets as well</code> - technically, it's: </p><ul><li>"the data is available on the 🤗 Hub", or</li><li>"the data is available on the Hugging Face Hub", or</li><li>"the data is available on the 🤗 Hub, and you can download it with 🤗 Datasets library."</li></ul>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (MKhalusova) [inline notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: <p>It would be better to define these constants closer to the training part of the tutorial. You don't use them until a few cells later. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (MKhalusova) [inline notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #6.    mean, std = [0.4914, 0.4822, 0.4465], [0.247, 0.243, 0.261]
</pre><p>Would be great to explain where you get these numbers from. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

Comment (MKhalusova) [inline notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: <p>I'm not sure we need to plot the distribution in this particular example. This is a curated dataset, of which we know for a fact that it is a balanced one. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='10'/>

Comment (MKhalusova) [inline notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #8.        num_workers=2*num_gpus,
</pre><p>IMO, instead of plotting the class distribution, it would be more helpful to explain some of the parameters, e.g. the choice of <code>num_workers</code>. Since you're using GPUs, it would also be helpful to state at the beginning of the notebook, what are the hardware requirements to run this notebook. </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='13'/>

**Comment by MKhalusova** [comment]
Thank you for working on this! Loved the style, and detailed explanations. I think we can skip the distribution plot in this example. Left a couple of comments in the notebook too.

File: notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb (hunk 1 above)
Comment (sitamgithub-MSIT) [reply notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: Apologies for the delayed response! I will fix the data visualization issue.

And for the data, as I am using the Torchvision dataset, there is no need to load it from the course asset. Once the reader executes the cell, it loads it in collab. To get to your point about importing data from the course repository, I am thinking about doing that in another notebook. This chapter is primarily concerned with practical applications. So, I thought to explore torchvision data, the hugging face dataset, and custom data (which may be Kaggle data) from the course repo. In this manner, I can thus reach every potential data loading point.

Comment (sitamgithub-MSIT) [reply notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: Regarding loading the data from Hugging Face Hub, as there will be multiple notebooks for this chapter, I thought to explore every option. Currently, I am considering custom data from the course repo, torchvision data, and hugging face data to be used in notebooks. For torchvision, I will keep the data source from torchvision data. When exploring the Timm ecosystem, hugging face hub data will be used.

Comment (sitamgithub-MSIT) [reply notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: Yeah. I will do that.

Comment (sitamgithub-MSIT) [reply notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: Yeah. Can be avoided, or I can add one line about that in the dataset description too.

Comment (sitamgithub-MSIT) [reply notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: Alright. Will take care of that.

Comment (MKhalusova) [reply notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: Is it explained anywhere else in the course? If so, you could link to that place. Otherwise, a link to a community post should be ok.

Comment (MKhalusova) [reply notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: Fair enough 👍

File: notebooks/Unit 2 - Convolutional Neural Networks/torchvision_Resnet.ipynb [hunk 2]
```diff
@@ -0,0 +1,1239 @@
+{
```
Comment (merveenoyan) [inline notebooks/Unit 2 - Convolutional Neural Networks/torchvision_Resnet.ipynb:1]: <p>Is it possible in this notebook you could add ResNet in transformers as well? <a href="https://huggingface.co/docs/transformers/v4.37.2/en/model_doc/resnet#overview" target="_blank">https://huggingface.co/docs/transformers/v4.37.2/en/model_doc/resnet#overview</a> It does abstract many things away compared to torchvision implementation. You can also make use of or demonstrate the new Backbone API: <a href="https://huggingface.co/docs/transformers/main/en/main_classes/backbones" target="_blank">https://huggingface.co/docs/transformers/main/en/main_classes/backbones</a></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

Comment (merveenoyan) [inline notebooks/Unit 2 - Convolutional Neural Networks/torchvision_Resnet.ipynb:1]: <p>I think this comment of @MKhalusova wasn't addressed (this cell should be closer to training)</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (merveenoyan) [inline notebooks/Unit 2 - Convolutional Neural Networks/torchvision_Resnet.ipynb:1]: <p>would be nice to have them side by side and smaller</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Review by merveenoyan — COMMENTED** [review]
thanks a lot for working on this 💜

**Comment by sitamgithub-MSIT** [comment]
> thanks a lot for working on this 💜

As both the torchvision part are almost ready and for vgg 19 reviewer comments are also addressed almost. So you can expect vgg19 to complete by eod. And for resnet also as the overall structure is same so shouldn't be any problem there as well.

**Comment by sitamgithub-MSIT** [comment]
@merveenoyan, except for two visualization problems, one is in plot distribution and another is in prediction. Everything I pushed into the latest commit. For those two viz changes, I need to modify the code and run the entire notebook again. If all the content is right, then I can make another final branch for vgg19 with those viz changes just for merge purposes, as most of the changes have already been made.

**Comment by merveenoyan** [comment]
it seems ready to merge to me, if @MKhalusova has no comments :)

File: notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb (hunk 1 above)
Comment (MKhalusova) [reply notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb]: <p>good point. </p>

File: notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb [hunk 3]
```diff
@@ -0,0 +1,1187 @@
+{
```
Comment (MKhalusova) [inline notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb:1]: <p>"In this notebook, We will learn" =&gt; "In this notebook, you will learn" </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='0'/>

File: notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb [hunk 4]
```diff
@@ -0,0 +1 @@
...9kJ33K9VjWWupKonDuK3MDK2S20YAaPSb8rFpa+rn2rVZsNfnbpq5VQCrKB4IKHKkFt/alJG0jY1L8y6qvb+WUGg3IzBQaaI25gFCpZ2j+mQpECVYNnNyrnyIkCMuwQaqaemovIg3Hrsxf0ihNsp97irPRyWbeTYfl8O2Vuf2mkfLL3tTHt6V4aNoBjSyj2pPXZQ+IMegpCA/RGDQ8MnqtxhCytnWs7bdxgfMLMyYp+yhS4RA5ZJUgZ8qFxlBzIXgaBSUYFxPaYX8syOZc50KiU7C0sr6IIJD1obUJoRaRANdpXttDGnskEoDav1RXZqLxdVvskS4fI3NLM65HDb80TosxUrzYzmboi4yGnxC6qSBKJgra23KPhCZ3O71bW8qc1sGHMxIWSuEohWC/M2ghJGDYrAVfYRMypxhoMTqy7qYc34p8qadqY/qR8kFpNdVWq38qrImGA+SaQ/yr7tmfBQ/HG0kSpCQAitE7GBIXkRqHsA8oACfAEKsNo1cDulvQZYEbxMsRK1EIGPAJZ26WXlNRHHTVSRRU0Nj8YEV9rACoOQguWX0TpjZGwB2rOAnKYySecWgtXNrtG4bAOBpJhsH5JDiPoQ4IILERB8aip6EL9nPwdQA7qSamvaaeFly3gwuoecEjwS3j4BrAZ4R4kCIySGyJIS1PIlxSJLQc6W+g6oZgQaSYlg0vVL2HGa8WrdqKwtr+83cP8b9VMGcmmvKh/IH0MAH3qPrPGYzh8XCA00Dbhpw38J1Ed0g4eUjM4JG2k1MSOyR45cr+OFIGi3Yw7uEbtZgPm+wmHfo5gKCGnbwwiJomdAsxN/IN5J7SLThjMYR2tZh3jVYzjus1gOGkLDuSec42ZiSzS6AXBSAjgByHeACnE8IsYGHR0MNGmrg4MWXnD184+HJwaey0cERSPcDDx5ixK8nDF9JWK8i1r1sfAHA0Kva6BzpvNDGBz/4Qdxyyy34nu/5HoQQ8LM/+7N41atehc997nPY2toCAPzET/wE/ut//a9417vehYMHD+L1r389fuiHfgj/+3//bwBAjBGvfvWrcfToUfyf//N/cP/99+Nv/I2/gbZt8Yu/+IvnU5yzUg16TDDa/LtJtjjV8qqds4VKRQi7QwdN8dOx81z9Rpm57Z4ifO1RiurQIn1hNPhU/swBCWRGIrHlZY8006SmEA2PhR+Mg0RcSkm
```
Comment (MKhalusova) [inline notebooks/Unit 2 - Convolutional Neural Networks/transfer_learning_vgg19.ipynb:1]: <p>Great that you tell the learner about the hardware requirements!</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/johko/computer-vision-course/pull/169/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='1'/>

**Review by MKhalusova — APPROVED** [review]
LGTM :)

**Review by johko — APPROVED** [review]

File: notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb (hunk 3 above)
Comment (sitamgithub-MSIT) [reply notebooks/Unit 2 - Convolutional Neural Networks/timm_Resnet.ipynb:1]: Done

**Comment by sitamgithub-MSIT** [comment]
@merveenoyan It is ready to be merged!

**Comment by johko** [comment]
Merging :rocket:
